File: backend/templates/workflow-engine/ai/extract.ts
/** AI extract handler — extracts structured fields from input. */

import { AINodeHandler } from './base';

export class ExtractHandler extends AINodeHandler {
  async execute(config: Record<string, unknown>): Promise<Record<string, unknown>> {
    const aiInput = this.resolver.resolveString((config.aiInput as string) ?? '');
    const aiFields = (config.aiExtractFields as string[]) ?? [];
    const aiPrompt = this.resolver.resolveString((config.aiPrompt as string) ?? '');

    const fieldsStr = aiFields.length > 0 ? aiFields.join(', ') : 'name, email, phone';

    const systemPrompt =
      'You are a data extraction assistant. Extract the following fields from the input: ' +
      `[${fieldsStr}]. ` +
      'Respond with JSON containing the extracted fields. Use null for fields not found.';
    const userPrompt = aiPrompt ? `${aiPrompt}\n\nInput:\n${aiInput}` : aiInput;

    const response = await this.callLLM(systemPrompt, userPrompt);
    const parsed = this.parseJsonResponse(response);

    return {
      extracted_fields: parsed,
      output: parsed,
    };
  }

  protected override mockResponse(): string {
    return JSON.stringify({ name: 'Mock User', email: 'mock@example.com' });
  }
}

File: backend/runtime/ai/extract.py
"""AI extract handler — extracts structured fields from input."""

import json
import logging

from runtime.ai.base import AINodeHandler

logger = logging.getLogger(__name__)


class ExtractHandler(AINodeHandler):
    """Extracts structured fields from input text."""

    async def execute(self, config: dict) -> dict:
        ai_input = self.resolver.resolve_string(config.get("aiInput", ""))
        ai_fields = config.get("aiExtractFields", [])
        ai_prompt = self.resolver.resolve_string(config.get("aiPrompt", ""))

        fields_str = ", ".join(ai_fields) if ai_fields else "name, email, phone"

        system_prompt = (
            "You are a data extraction assistant. Extract the following fields from the input: "
            f"[{fields_str}]. "
            "Respond with JSON containing the extracted fields. Use null for fields not found."
        )
        user_prompt = f"{ai_prompt}\n\nInput:\n{ai_input}" if ai_prompt else ai_input

        response = await self._call_llm(system_prompt, user_prompt)
        parsed = self._parse_json_response(response)

        return {
            "extracted_fields": parsed,
            "output": parsed,
        }

    def _mock_response(self, system_prompt: str, user_prompt: str) -> str:
        return json.dumps({"name": "Mock User", "email": "mock@example.com"})
